# Part 1. Equation of a Slime

How many late days are you using for this assignment? zero

In [73]:
# Imports section
%pip install scikit-learn
%pip install pandas
%pip install numpy
import pandas as pd
import sklearn as sk
import numpy as np


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


## 1. Loading the dataset

In [74]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

science_data = pd.read_csv("science_data_large.csv")

display(science_data.head(15))
science_data.describe()


,Temperature °C,Mols KCL,Size nm^3
0,469,647,6.244743e+05
1,403,694,5.779610e+05
2,302,975,6.196847e+05
3,779,916,1.460449e+06
4,901,18,4.325726e+04
5,545,637,7.124634e+05
6,660,519,7.006960e+05
7,143,869,2.718260e+05
8,89,461,8.919803e+04
9,294,776,4.770210e+05


,Temperature °C,Mols KCL,Size nm^3
count,1000.000000,1000.000000,1.000000e+03
mean,500.500000,471.530000,5.086111e+05
std,288.819436,288.482872,4.474838e+05
min,1.000000,1.000000,1.611429e+01
25%,250.750000,226.750000,1.298267e+05
50%,500.500000,459.500000,3.827182e+05
75%,750.250000,710.250000,7.603211e+05
max,1000.000000,1000.000000,1.972127e+06


## 2. Splitting the dataset

In [75]:
# Take the pandas dataset and split it into our features (X) and label (y)

X = science_data[['Temperature °C', 'Mols KCL']]
y = science_data['Size nm^3']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42

X_train, X_test, y_train, y_test = sk.model_selection.train_test_split(X, y, test_size=0.1, random_state=42)

## 3. Perform a Linear Regression

In [76]:
from sklearn.linear_model import LinearRegression

# Use sklearn to train a model on the training set
reg = LinearRegression()
reg.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample_data = pd.DataFrame({'Temperature °C': [450], 'Mols KCL': [500]})
prediction = reg.predict(sample_data)
print("Size prediction based off sample data:", prediction[0])

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
score = reg.score(X_test, y_test)
print("Score of model:", score)

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
coeff = reg.coef_
intercept = reg.intercept_
print("Coefficient:", coeff)
print("Intercept:", intercept)


Size prediction based off sample data: 496721.939676855
Score of model: 0.8552472077276095
Coefficient: [ 866.14641337 1032.69506649]
Intercept: -409391.47958340764


Write the linear equation of a slime: (example equation: $E = mc^2$)

The equation for the model $h(x)$ is:

$$
h(x) = 866.14641337 \cdot \text{Temperature} + 1032.69506649 \cdot \text{Mols KCL} - 409391.47958340764
$$



Report on score and explain meaning:

The score of the linear regression model is 0.8552472077276095. This score is the coefficient of determination and it tells how well the model explains the variability of the label based on the features. If the score is 1 it means the model perfectly predicts the target variable and if it is 0 is means the model does not explain the variability of the label.

The score given by our model of .855 means that about 85.5% of the variability of the size, nm^3, can be explained by the model using the two features, Temperature and Mols KCL. Since .855 is fairly close to to 1.0 it means the model is doing a good job of describing the relationships in the data and is making pretty accurate predictions on new data, like the test data.

## 4. Use Cross Validation

In [77]:
from sklearn.model_selection import cross_val_score, ShuffleSplit
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42

n_samples = X.shape[0]
cv = ShuffleSplit(n_splits=5, test_size=0.1, random_state=42)
cv_scores = cross_val_score(reg, X, y, cv=cv)

# Report on their finding and their significance
print("Cross-validated scores for each fold:", cv_scores)
print("Mean cross-validated score:", cv_scores.mean())

Cross-validated scores for each fold: [0.85524721 0.85830729 0.8654508  0.83855498 0.83002469]
Mean cross-validated score: 0.8495169947353961


Write findings here:

The scores across 5 folds are: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101] and the mean score is: 0.8597294202684646.
This mean score is slightly higher than the score found above with just the regular linear regression. Since the score is high and close to 1.0 it means the model will most likely perform well on unseen data meaning it is good at generalization and is not to overfit to the training data. Showing the model has done well to capture underlying patterns in the data.

## 5. Using Polynomial Regression

In [78]:
from sklearn.preprocessing import PolynomialFeatures
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)

poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)

cv_scores_poly = cross_val_score(reg, X_poly,y, cv=cv)

reg.fit(X_poly, y)

print("Cross-validated scores for each fold (Polynomial Regression):", cv_scores_poly)
print("Mean cross-validated score (Polynomial Regression):", cv_scores_poly.mean())

coefficients_poly = reg.coef_
intercept_poly = reg.intercept_
print("Coefficient poly:", coefficients_poly)
print("Intercept poly:", intercept_poly)

# Report on the metrics and output the resultant equation as you did in Part 3.

Cross-validated scores for each fold (Polynomial Regression): [1. 1. 1. 1. 1.]
Mean cross-validated score (Polynomial Regression): 1.0
Coefficient poly: [ 0.00000000e+00  1.20000000e+01 -1.23110940e-07 -1.05650504e-11
  2.00000000e+00  2.85714287e-02]
Intercept poly: 1.657102257013321e-05


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

$$
h(x_1,x_2) = 1.6571 \times 10^{-5} + 12 x_1 + (−1.231 \times 10^{−7}) x_2 + (-1.0565 \times 10^{-11}) x_1^2 + 2 x_1 x_2 + 2.857 \times 10^{-2} x_2^2
$$

Report on the score and interpret:

The cross validated scores are [1. 1. 1. 1. 1.] with a mean of 1.0. This means that in each of the 5 CV folds the model had a perfect fit and explained 100% of the variance of the label, size, based on the features, temperature and Mols KCL, so for each fold the model was able to perfectly predict the output values. Since all scores from each fold are the same the standard deviation and variance are zero meaning the model is stable and reliable.

The score being a perfect 1.0 implies that there is a possibility for overfitting because of data leakage between the train and test splits. The prefect score could also imply that the model is too complex for the amount of data available or that since this is synthetic dataset it is able to find an exact relationship.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [79]:
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
# Load the dataset. Then train and evaluate the classification models.

random_seed = 42

ckd_data = pd.read_csv("ckd_feature_subset.csv")

display(ckd_data.head())


X_ckd = ckd_data.drop(columns=["Target_ckd"])
y_ckd = ckd_data["Target_ckd"]

X_ckd_train, X_ckd_test, y_ckd_train, y_ckd_test = sk.model_selection.train_test_split(X_ckd, y_ckd, test_size=0.1, random_state=42)

,age,bp,wbcc,appet_poor,appet_good,rbcc,Target_ckd
0,0.688312,0.333333,0.000000,1,0,0.000000,1
1,0.545455,0.333333,0.128319,1,0,0.305085,1
2,0.714286,0.500000,0.238938,1,0,0.186441,1
3,0.688312,0.333333,0.283186,0,1,0.338983,1
4,0.441558,0.333333,0.221239,1,0,0.220339,1


In [80]:
from sklearn.linear_model import LogisticRegression

models = {
    "Logistic Regression": LogisticRegression(),
    "SVM": SVC(),
    "KNN": KNeighborsClassifier(),
    "Neural Network": MLPClassifier()
}

n_samples = X_ckd.shape[0]
cv = ShuffleSplit(n_splits=5, test_size=0.1, random_state=random_seed)

model_results = {}
for name, model in models.items():
    model.fit(X_ckd_train, y_ckd_train)
    scores = cross_val_score(model, X_ckd, y_ckd, cv=cv)
    model_results[name] = {
        'Average Accuracy': np.average(scores),
        'Standard Deviation': np.std(scores)
    }


results_df = pd.DataFrame(model_results).T
print("Model Performance (5-Fold Cross-Validation):")
display(results_df)


c:\Users\Aflam\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\Aflam\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\Aflam\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\Aflam\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the opti

Model Performance (5-Fold Cross-Validation):


c:\Users\Aflam\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


,Average Accuracy,Standard Deviation
Logistic Regression,0.8750,0.111803
SVM,0.9625,0.050000
KNN,0.9750,0.030619
Neural Network,0.9000,0.108972


## Results and Conclusion for Classification Experiments

KNN performed the best of all the models with the highest average accuracy of 97.5% and the lowest standard deviation of .031. This high accuracy shows that KNN classified the highest proportion of samples and had consistent performance across the different folds of cross-validation shown by its low std. KNN most likely performed well because of the simplicity of its algorithm which makes predictions based on the majority class among the nearest neighbors.

SVM also performed well most likely because of its ability to handle high-dimensional spaces and find complex decision boundaries.

Neural Network performed a little worse than SVM and Logistic Regression performed the worst most likely because it is linear it might struggle with complex non-linear relationships in the data.

In [81]:
# Experiments with Neural Network.

# Experiment with Neural Network configurations
nn_configurations = {
    'Default Configuration': MLPClassifier(hidden_layer_sizes=(100,), max_iter=1000, random_state=random_seed),
    'More Hidden Layers': MLPClassifier(hidden_layer_sizes=(100, 50), max_iter=1000, random_state=random_seed),
    'Different Learning Rate': MLPClassifier(hidden_layer_sizes=(100,), learning_rate_init=0.01, max_iter=1000, random_state=random_seed)
}

nn_results = {}
for config_name, model in nn_configurations.items():
    scores = cross_val_score(model, X_ckd, y_ckd, cv=cv)
    nn_results[config_name] = {
        'Average Accuracy': scores.mean(),
        'Standard Deviation': scores.std()
    }

# Create a results DataFrame for Neural Networks
nn_results_df = pd.DataFrame(nn_results).T
display(nn_results_df)


,Average Accuracy,Standard Deviation
Default Configuration,0.9625,0.050
More Hidden Layers,0.9875,0.025
Different Learning Rate,1.0000,0.000


## Results and Conclusion for Neural Network Experiments

The Default Configuration with a single hidden layer had an average accuracy of 96.25% with a std of .05 which is a pretty good performance.

The second configuration with an additional hidden layer had an increase in performance to 98.75% and a decrease in std to .025 showing that adding more hidden layers allowed the model to learn more complex relations in the data increasing performance.

The final configuration with a different learning rate of .01 was able to fit the data perfectly showing that learning rate is very important in improving the models accuracy. The perfect score could potentially hint at an over complexity in this configuration and some overfitting to the data occurring.

From the testing the number of hidden layers and learning rate of a large affect on the accuracy and performance of the neural network model. More hidden layers allow the model allows the model to pick up on more complex relationships in the data. Adjusting the learning rate can lead to quick convergence and better results. It is important to keep a balance in the complexity to prevent the risk overfitting and maintaining a model that has good generalization capabilities.